### PR TITLE
Some document style fixes.

### PR DESCRIPTION
**Pull request by peter50216**
Make docs style consistent.
Also add a tag `@diff` for difference with python-pwntools.
Also refactor MemLeak / DynELF and add docs for methods.

---
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/22)

### CODE REVIEW

File: lib/pwnlib/constants/constants.rb [hunk 1]
```diff
@@ -1,20 +1,21 @@
...
   # @example
   #   context.arch = 'amd64'
   #   Pwnlib::Constants.SYS_read
   #   # => Constant('SYS_read', 0x0)
   module Constants
     # @note Do not create and call instance method here. Instead, call module method on {Constants}.
     module ClassMethods
-      include ::Pwnlib::Context
-      # Try getting constants when method missing
+      # To support getting constants like Pwnlib::Constants.SYS_read.
```
Comment (david942j) [inline lib/pwnlib/constants/constants.rb]: +Pwnlib::Constants.SYS_read+ ?

File: lib/pwnlib/dynelf.rb [hunk 2]
```diff
@@ -19,42 +18,86 @@ class DynELF
 
     attr_reader :libbase
 
+    # @param [Integer] addr
+    #   An address known to be inside the ELF.
+    #
+    # @yieldparam [Integer] leak_addr
+    #   The start address that the leaker should leak from.
```
Comment (david942j) [inline lib/pwnlib/dynelf.rb]: one more empty line here?

File: lib/pwnlib/dynelf.rb [hunk 3]
```diff
@@ -19,42 +18,86 @@ class DynELF
...
-      @hshtab ||= find_dt(DT_GNU_HASH)
-      @strtab ||= find_dt(DT_STRTAB)
-      @symtab ||= find_dt(DT_SYMTAB)
-      resolve_symbol_gnu(symb)
+    # Lookup a symbol from the ELF.
+    #
+    # @param [String] symbol
+    #   The symbol name.
+    #
+    # @return [Integer]
```
Comment (david942j) [inline lib/pwnlib/dynelf.rb]: `[Integer, nil]`

File: lib/pwnlib/util/fiddling.rb [hunk 4]
```diff
@@ -166,11 +171,11 @@ def bits_str(s, endian: 'big', zero: 0, one: 1)
         #
         # @param [String, Array<String, Integer, Boolean>] s
         #   String or array of bits to be convert back to string.
-        #   <tt>[0, '0', false]</tt> represents 0-bit,
-        #   and <tt>[1, '1', true]</tt> represents 1-bit.
+        #   <tt>[0, '0', false]</tt> represents 0-bit, and <tt>[1, '1', true]</tt> represents 1-bit.
```
Comment (david942j) [inline lib/pwnlib/util/fiddling.rb]: suddenly の `<tt>`

Comment (peter50216) [inline lib/pwnlib/util/fiddling.rb]: Because yard parsing is weird... +...+ doesn't work here.

File: lib/pwnlib/util/packing.rb [hunk 5]
```diff
@@ -14,270 +13,267 @@ module Util
...
+      #   Endian to use when unpacking.
+      #   Can be any value accepted by context (See {Context::ContextType}).
+      #   Default to +context.endian+.
+      # @param [Boolean, String] signed
+      #   Whether the output number should be signed.
+      #   Can be any value accepted by context (See {Context::ContextType}).
+      #   Default to +context.signed+.
+      #
+      # @return [Integer]
+      #   The unpacked number.
```
Comment (david942j) [inline lib/pwnlib/util/packing.rb]: One more blank line here?

File: STYLE.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+b.downto(a) { ... }
+
+# Bad
+(a..b).each { ... }
+(a..b).reverse_each { ... }
+a.upto(b).map { ... }
+```
+
+## YARD tag order
+Tag order is `@param, @return, @yieldparam, @yieldreturn, @todo, @note, @diff, @example`.
```
Comment (david942j) [inline STYLE.md]: miss `@raise`

File: STYLE.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+A new line between each group,
+and sort alphabetically in each group.
+
+## Document style
+* Public methods intended to be used by user should be documented.
+* `@param`, `@return` document on a indented new line.
+```ruby
+# Good
+# @param [String] bla
+#   a bla argument.
```
Comment (david942j) [inline STYLE.md]: `#   A bla argument.`

File: STYLE.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+# Bad
+# @param [String] bla a bla argument.
```
Comment (david942j) [inline STYLE.md]: `A bla argument.`

**Comment by david942j** [comment]
Seems Travis CI didn't test this PR?

**Comment by peter50216** [comment]
Review status: 0 of 27 files reviewed at latest revision, 6 unresolved discussions.

---

*[STYLE.md, line 24 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-KgDWfOl0b3R1mZUuyj6-r4-24:-Kh1bSOEPp3nUeDDSGHe:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/2ae7f045db3d01b7b5a7aee704201e72c3313bd3/STYLE.md#L24)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`@raise` missing
</blockquote></details>

Done.

---

*[STYLE.md, line 42 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-KgDWfOl0b3R1mZUuyj6-r4-42:-Kh1bTq4Glb0LFGWnsdD:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/2ae7f045db3d01b7b5a7aee704201e72c3313bd3/STYLE.md#L42)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`#   A bla argument.`
</blockquote></details>

Done.

---

*[STYLE.md, line 45 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-KgDWfOl0b3R1mZUuyj6-r4-45:-Kh1bUXQmP4Zg5DnufwE:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/2ae7f045db3d01b7b5a7aee704201e72c3313bd3/STYLE.md#L45)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`A bla argument.`
</blockquote></details>

Done.

---

*[lib/pwnlib/dynelf.rb, line 42 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-KfzGK3sReXm4X_juEsh-r1-42:-Kh1bVgf2a9cCVsJEqI9:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/7f10b8a35bafaa4491301f1fa3742d241011255d/lib/pwnlib/dynelf.rb#L42)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`[Integer, nil]`
</blockquote></details>

Done.

---

*[lib/pwnlib/constants/constants.rb, line 18 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-KfzGK3sReXm4X_juEsf-r1-18:-Kh1bWWZVzZO5HZn85n1:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/7f10b8a35bafaa4491301f1fa3742d241011255d/lib/pwnlib/constants/constants.rb#L18)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

+Pwnlib::Constants.SYS_read+ ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22)*

**Comment by david942j** [comment]
Review status: 0 of 27 files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/pwnlib/dynelf.rb, line 40 at r5](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-Kh21PoQbVp_ztkDn0dV:-Kh21PoQbVp_ztkDn0dW:brzuhzh) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/c072e0b2224bfe26eb8d1c5daa22cddad80fa3ae/lib/pwnlib/dynelf.rb#L40)):*
> ```Ruby
>     # Lookup a symbol from the ELF.
>     #
>     # @param [String] symbol
> ```

Support [String, Symbol] (?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22)*

**Comment by david942j** [comment]
Reviewed 7 of 17 files at r1, 9 of 17 files at r2, 1 of 4 files at r3, 8 of 9 files at r4, 1 of 1 files at r5.
Review status: 26 of 27 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22)*

File: test/util/packing_test.rb [hunk 9]
```diff
@@ -4,7 +4,7 @@
 require 'pwnlib/util/packing'
```
Comment (david942j) [inline test/util/packing_test.rb]: sort ?

**Comment by david942j** [comment]
Review status: 26 of 27 files reviewed at latest revision, 2 unresolved discussions.

---

*[lib/pwnlib/dynelf.rb, line 40 at r5](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-Kh21PoQbVp_ztkDn0dV:-Kh26FBscE77jpTNjcdB:b5s57xm) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/c072e0b2224bfe26eb8d1c5daa22cddad80fa3ae/lib/pwnlib/dynelf.rb#L40)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

Support [String, Symbol] (?)
</blockquote></details>

OK because should not in this PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22)*

**Comment by peter50216** [comment]
Review status: 16 of 30 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/util/packing_test.rb, line 4 at r5](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-KgCn_RioYGYa4-pm02V-r5-4:-Kh5rvisKb5mw2hejLPY:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/c072e0b2224bfe26eb8d1c5daa22cddad80fa3ae/test/util/packing_test.rb#L4)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

sort ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22)*

**Comment by david942j** [comment]
:lgtm:

---

Reviewed 1 of 17 files at r2, 13 of 13 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/22#-:-Kh6KdiWdEUGHouV3b07:bnfp4nl)*

**Pull request merged**